### ISSUE TITLE
[HOLD for payment 2023-12-26] [$1000] mWeb - Composer - Unnecessary blank area is created when scrolling down while keyboard is open

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#19642]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the testing link https://staging.new.expensify.com/
2. Open any chat
3. Write multiline text in the composer until the enlarge composer icon appear (keep the keyboard open)
4. Tap on the maximize composer icon
5. Above the keyboard, scroll down

## Expected Result:
The bottom of the screen should stick to the bottom of the input composer, and no extra blank area should be created when the user scrolls down on the screen

## Actual Result:
When scrolling down while the composer is maximized, an excessively wide area is created below the composer, resulting in continuous resizing of the composer when scrolling both upward and downward

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.18.0

**Reproducible in staging?:** yes

**Reproducible in production?:** Yes - but the composer does not keep resizing, only the blank area is presented

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://platform.applause.com/services/links/v1/external/8e4abb883a4188a43758f0561fbfd5c7900892adaded42ecb9fb365ac7c0b0a3

![](https://platform.applause.com/services/links/v1/external/c18278821c88511861507f35475e32e2c94c02f87ff33f8902930ca384755d44)

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01302d5f678fb0c5f7</li>
        <li>Upwork Job ID: 1662142490327298048</li>
        <li>Last Price Increase: 2023-05-26</li>
<li>Automatic offers: </li>
<ul>
<li>suneox | Contributor | 28039355</li>
</ul></li>
    </ul>
</details>

**Comment by mountiny** [comment]
We can export this to Callstack

**Comment by lukemorawski** [comment]
here I am

**Comment by qasimkhan23** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When scrolling down while the composer is maximized, an excessively wide area is created below the composer, resulting in continuous resizing of the composer when scrolling both upward and downward.
### What is the root cause of that problem?
We have added the marginTop property in screenWrapperStyle and we are providing the value to it from viewportOffsetTop which is causing the marginTop value to fluctuate when a user drag the screen in IOS Safari browser, and the View which contains the chat has given the property flex:1 because of that the Inner View adjusts its size and the composer looks resizing continuously.
https://github.com/Expensify/App/blob/967100432a5f3ae5458f43ab20163ea07f6d17dc/src/pages/home/ReportScreen.js#L229

before 
<img width="361" alt="Screenshot 2023-05-29 at 4 01 20 PM" src="https://github.com/Expensify/App/assets/46372909/416db9e3-93cb-4b07-902c-4bff659ea8d2">
you can see below, how the top of the page got cut when I scroll the page down, when I scroll the page down the offset value is provided in a negative number which results in cutting the top part of the page because of marginTop: negativeValue
<img width="354" alt="Screenshot 2023-05-29 at 4 11 41 PM" src="https://github.com/Expensify/App/assets/46372909/736cb72b-681d-4f99-a45a-551ccf75b6d8">


### What changes do you think we should make in order to solve the problem?
There is no need to add viewportOffsetTop because we have already provided flex:1 to wrapper which will take care of different devices.
after applying the solution 
<img width="355" alt="Screenshot 2023-05-29 at 4 04 56 PM" src="https://github.com/Expensify/App/assets/46372909/6d4cb9e0-66e4-423c-9a47-66e231859ec0">

```
  const screenWrapperStyle = [styles.appContent, styles.flex1];

```

### What alternative solutions did you explore? (Optional)
None

**Comment by Santhosh-Sellavel** [comment]
This is picked by call stack waiting for @lukemorawski

**Comment by lukemorawski** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The problem we are trying to solve is the incorrect behavior of the input composer in the chat feature of the Expensify web application on iOS/Safari. When the composer is focused and the user scrolls down, an excessively wide area is created below the composer, resulting in continuous resizing of the composer when scrolling in both directions. This behavior is not aligned with the expected behavior, which is for the input composer to stick to the bottom of the screen without any extra blank area being created when scrolling. This behaviour occurs not only for a maximised composer but also for the compact one. 

### What is the root cause of that problem?

Root cause of the issue seems to be related to the calculation of the height of the root div and the scrolling behavior. The calculation might not be taking into account the correct dimensions and resulting in an incorrect resizing behavior. On iOS Safari the view port height does change (decreases) when a virtual keyboard appears but the root div height seems to stay the same, allowing the user to scroll the content, in including the composer, below the virtual keyboard. 

### What changes do you think we should make in order to solve the problem?

The solution that is most compatible with the current code base, the way the UI elements are laid out, and constraints of RNW is to hide the virtual keyboard when a user manages to scroll the view port (not the div where the messages are)

### What alternative solutions did you explore? (Optional)

An alternative to that would be to change the the way the elements are laid out on the mobile browser and the way element heights are calculated using browser specific APIs departing from default RNW solutions.

**Comment by lukemorawski** [comment]
Working on a solution. Already have one but investigating a more comprehensive one.

**Comment by lukemorawski** [comment]
Still working on a more robust solution. Will have more info today.

**Comment by lukemorawski** [comment]
After some trial and error I can say that preventing scrolling of the whole window when the user activates the virtual keyboard is not achievable for current app UI setup. Common solutions and libraries (such as `body-scroll-lock`) don't seem to have an effect.
The whole thing is caused by iOS Safari rather peculiar way of dealing with a problem of obscured input elements. To prevent any input element from being hidden under the keyboard, in an eventuality when there is not enough room under that element, Safari adds some additional space at the bottom of the window. The whole problem is that this space seems to be not a part of the DOM tree. It's not added body or html height or on any other node. It's just some space under the DOM tree. This means that it's beyond control of the available CSS or JS apis. The only way to mitigate this problem is to prevent the window element from scrolling, by applying some styles to body and preventing default behaviour of the scroll event (plus some other more finesse tricks). Unfortunately due to a very deep anchoring of the scrolled div and it's layout this doesn't work.
The only way I have found is to scroll the window content back to it's optimal position after the user over scrolled it downwards.
I have a working solution but this needs to be discussed internally and now I'm delegated to an another, more important issue, so the PR will be slightly delayed.

**Comment by Santhosh-Sellavel** [comment]
@lukemorawski 

Can you follow the proposal [template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md) and leave a new comment

**Comment by Santhosh-Sellavel** [comment]
@lukemorawski Any update?

**Comment by lukemorawski** [comment]
@Santhosh-Sellavel Posted internally an updated proposal with a working solution. Waiting for acceptance.

**Comment by lukemorawski** [comment]
## Updated Proposal

### Please re-state the problem that we are trying to solve in this issue.

The problem we are trying to solve is the incorrect behavior of the input composer in the chat feature of the Expensify web application on iOS/Safari. When the composer is focused and the user scrolls down, an excessively wide area is created below the composer, resulting in continuous resizing of the composer when scrolling in both directions. This behavior is not aligned with the expected behavior, which is for the input composer to stick to the bottom of the screen without any extra blank area being created when scrolling. This behaviour occurs not only for a maximised composer but also for the compact one.

### What is the root cause of that problem?

The whole thing is caused by iOS Safari rather peculiar way of dealing with a problem of obscured input elements. To prevent any input element from being hidden under the keyboard, in an eventuality when there is not enough room under that element, Safari adds some additional space at the bottom of the window. The whole problem is that this space seems to be not a part of the DOM tree. It's not added body or html height or on any other node. It's just some space under the DOM tree. This means that it's beyond control of the available CSS or JS apis. The DOM api registers `viewPort` and `window` height decrease after the keyboard shows up, yet in fact you can see that this space increases. The only way to mitigate this problem is to prevent the window element from scrolling, by applying some styles to body and preventing default behaviour of the scroll event (plus some other more finesse tricks). Unfortunately due to a very deep anchoring of the scrolled div and it's layout this doesn't work.

### What changes do you think we should make in order to solve the problem?

After some trial and error I can say that preventing scrolling of the whole window when the user activates the virtual keyboard is not achievable for current app UI setup. Common solutions and libraries (such as body-scroll-lock) don't seem to have an effect.
The solution I have found to work is to scroll back the window content back to it's optimal position after the user stops scrolling (see attached video)
This is can be done by creating a HOC that wraps the ReportScreen and checks if the keyboard is open, if so then scrolls the window back to it's desired position after the `touchend` event.
This also requires adding a helper module that would extend RN Keyboard to use `addEventListener` on the web. Currently RNW exposes this method, but it does nothing. We could add a module that would simply reexport all the methods on native, while on the web, it would add this missing method using browser APIs (checking if the VP size changed, and then checking for active element and if it's an element that exposes a keyboard).
Change to `SwipeableView` would be also desirable to make to `onSwipeDown` callback work on the web (currently only native). This would prevent unnecessary scrolling of the window while swiping on the composer div. This can be done by listening to the `touchstart`/`touchend` events and calculating the diff between the `clientY` values. When that diff is greater then some minimal value, the `onSwipeDown` callback is triggered.

### What alternative solutions did you explore? (Optional)

Scroll blocking was explored but only seemed to work for elements directly inside the body element.


https://github.com/Expensify/App/assets/134388952/c67523c4-c6cc-4f26-92ff-43c692110770

> Works also when the composer text area field is expanded

**Comment by abekkala** [comment]
@Santhosh-Sellavel what do you think of the updated proposal above

**Comment by lukemorawski** [comment]
@Santhosh-Sellavel Well maybe in this case it's worth having a workaround and leave the end user with an incrementally better UX, than just leave it as it is, because the proper solution is unachievable due to a browser bug/odd way of dealing with a simple thing?

**Comment by thienlnam** [comment]
Hmm, is this something that would eventually be fixed within iOS Safari and wouldn't this apply to all cases where you're able to scroll and the keyboard is open on iOS? 

We've definitely added platform specific code to handle certain cases and then it only caused us headache in the future having to deal with other edge cases that ended up popping up from those changes. This kind of feels like the same scenario that might happen. 

I'm more curious if there is an open bug report for this on iOS and if we can apply pressure to have it fixed there

**Comment by lukemorawski** [comment]
@thienlnam It is reported as a bug here:
https://bugs.webkit.org/show_bug.cgi?id=141832
Yet, looks like authors treat it more like a feature :) It was reported in 2015 and till this day it wasn't resolved, and probably won't be. 
There might be some how in new `dvh` view port height unit, but it's only supported in the very latest browsers and it needs to be tested, specifically in mobile safari. WebKit might still add this extra space under the vp, no new length unit won't help.

**Comment by thienlnam** [comment]
Ah yeah yikes - this isn't ideal. Shall we try to this change and see how it holds up?

**Comment by lukemorawski** [comment]
@thienlnam I would certainly give it a try. It's easy to remove in the future by simply deleting a HOC from the list in `compose`.

**Comment by lukemorawski** [comment]
@thienlnam @Santhosh-Sellavel any news on that one?

**Comment by Santhosh-Sellavel** [comment]
Yeah, let's give it a try. Please work on it.

**Comment by lukemorawski** [comment]
Great!

**Comment by abekkala** [comment]
@JmillsExpensify Reassigning for Sabbatical 
____
### CURRENT STATUS:

This will be heading to payment stage soon. 

* Issue reported by:  **Applause - Internal Team**
* Contributor selected for fix: @lukemorawski (Callstack)
* C+ Review: @Santhosh-Sellavel

**Comment by thienlnam** [comment]
Issue is in the reviewing stage - Luke has been out for a bit and just recently got back so we're still chugging along

**Comment by JmillsExpensify** [comment]
Thanks for updating!

**Comment by thienlnam** [comment]
After some recent changes in main, there is some unexpected functionality that is blocking the PR 
https://github.com/Expensify/App/pull/21298#issuecomment-1680572874

We're going to hold this ticket, and create another ticket for this new bug that is on main

**Comment by thienlnam** [comment]
Cc @lukemorawski Could you summarize what the bug is and steps to reproduce so we can get a bug created?

Essentially this format
BUG: 

Action Performed:

Expected Result: 
Actual Result:
Platforms: 
Version Number:
Screenshots or Videos: :thread:

**Comment by JmillsExpensify** [comment]
What's the latest on this issue?

**Comment by thienlnam** [comment]
Looks like the bug was fixed in ios17, and cannot be reproduced - I believe we should be able to merge main and continue in this issue then. @lukemorawski do you also agree?

**Comment by JmillsExpensify** [comment]
Sounds great to me! Let's get this one wrapped up.

**Comment by lukemorawski** [comment]
@thienlnam @JmillsExpensify yep, sounds good

**Comment by JmillsExpensify** [comment]
PR review on-going

**Comment by JmillsExpensify** [comment]
@parasharrajat Mind kicking us off with the checklist?

**Comment by JmillsExpensify** [comment]
In the meantime, here's the payment summary:

 - Issue reporter: N/A
 - Contributor: No payment required
 - Contributor+: @parasharrajat $1,000

**Comment by lukemorawski** [comment]
@parasharrajat Not sure when I'll be able to get my hands on this. I have several high priority tickets to do. Possibly next week.

**Comment by JmillsExpensify** [comment]
Still working through this issue.

**Comment by cead22** [comment]
No objections from me, but I'll let @thienlnam assign since he has a lot more context around this issue than me

**Comment by thienlnam** [comment]
> ios safari will extend a space by VisualViewport

I'm not quite following this root case, what does this mean? 

> On [ReportActionsList](https://github.com/Expensify/App/blob/aa42ddc221f8fab2e470944549fab835eef42f00/src/pages/home/report/ReportActionsList.js#L457) has added styles.overscrollBehaviorContain to avoid scrolling over parent so the extended space from VisualViewport will not shown

So is the style being added only to prevent the VisualViewport from being shown on iOS safari? 

I am not totally sure the implications on these changes, seems like something we can try out. Can you find a link to the bug report in webkit so that we can link to it and remove these safari specific changes when the bug is resolved in future versions?

**Comment by thienlnam** [comment]
Are you able to search for the bug within this [list](https://bugs.webkit.org/buglist.cgi?quicksearch=viewport)? 

Other than that, I think we're good to give this a shot - I'll go ahead and assign you

**Comment by JmillsExpensify** [comment]
PR up for review

**Comment by JmillsExpensify** [comment]
Hmmm, I'm a bit confused where we're at on payment. It looks like the original contributor's PR was reverted, so that leaves us with @suneox who was assigned more recently. That payment summary would be:

 - Issue reporter: QA
 - Contributor: @suneox $500
 - Contributor+: @parasharrajat $500

**Comment by JmillsExpensify** [comment]
I think it's incorrect automated logic on our side. Given that you were assigned a couple of weeks ago, I think the $500 offer is correct and I can pay that amount on payout. Do you agree?

**Comment by thienlnam** [comment]
@JmillsExpensify This bug is very old and has retained the $1000 price tag. I think since this issue predates all our pricing changes we should just pay this one out at the original price

**Comment by JmillsExpensify** [comment]
Cool, I'm not incredibly passionate. Updated the payment summary above.

**Comment by JmillsExpensify** [comment]
Contributor has been paid out and regression created. @parasharrajat Please submit a request via Expensify and we'll get your paid.

**Comment by JmillsExpensify** [comment]
@laurenreidexpensify Do you mind confirming the payment summary on this issue before I approve payment?

**Comment by laurenreidexpensify** [comment]
Payment Summary: 

$1000 Contributor - @suneox payment issued in Upwork 
$1000 C+ - @parasharrajat requested in newdot

**Comment by JmillsExpensify** [comment]
$1,000 approved for @parasharrajat
